### ISSUE TITLE
File Duplication Bug In Outputted Zips

### ISSUE THREAD

**Issue by AverniteDF** [WinMerge/winmerge#1588]
I seem to be encountering a bug when doing the following in WinMerge:

Ctrl + A to select all files
Right Click > Zip > Left to...

This is a great feature for sending all changes to a zip archive. However, some of the files are duplicated in the generated archive:

![1](https://user-images.githubusercontent.com/115361770/203885765-9d4858b6-16ef-46d2-84de-49ddd954df26.png)
In the Mods folder, everything looks fine

![2](https://user-images.githubusercontent.com/115361770/203885810-3be23a44-54f1-4b7b-b91e-8db95b81bbaa.png)
Moving into the Zoom subfolder, all files are duplicated

![3](https://user-images.githubusercontent.com/115361770/203885845-6a93d419-1a2f-47d7-b453-6143e7c9a7ed.png)
Moving deeper into the Scripts subfolder, all files are triplicated!

I encountered this problem in 2.16.16.0 so I downloaded the latest version 2.16.24.0 and this issue is still present.

Here is the configuration log:
[WinMerge.txt](https://github.com/WinMerge/winmerge/files/10088106/WinMerge.txt)

**Comment by sdottaka** [comment]
Thank you for your report.
I have found that the problem is reproduced when folders are expanded.

The workaround for this problem is to select the [View]->[Collapse All Subfolders] menu item before creating the compressed file.

**Comment by AverniteDF** [comment]
Thank you for the timely response!

Using Tree View with collapsed subfolders eliminates the file duplication
issue.

However, the resulting zip archive is now much larger (in my case 20 MB vs
2 MB) and contains unwanted files (ones that don't have differences from
the right side).

Maybe there's a quick fix that could be implemented by adding each file
path to a hashset and checking for duplicates before adding to zip (or
something like that)?

**Comment by AverniteDF** [comment]
I believe a viable workaround is to only select the listed files and none
of the folders before doing Right-Click > Zip.

As for a solution: When not in tree mode, WinMerge should never add the
contents of folders to a zip. The folder itself may be added to the zip if
no other selected files contain that folder in their path. If you don't
want the extra overhead of performing that check, then simply add the
folder (but not its contents) as I don't think this will adversely affect
the zip.

**Comment by AverniteDF** [comment]
It looks like whenever a folder item is passed to 7z it naively adds the entire contents of that folder to the compressed archive, which in many cases is not what you want. WinMerge itself should decide which files to add.

I wasn't able to find the spot in the code where this 'handing off' to 7z happens, but the following modification provided a partial solution:

**7zCommon.cpp**

```
Merge7z::Envelope *DirItemEnumerator::Enum(Item &item)
{
	const CDiffContext& ctxt = m_pView->GetDiffContext();
	const DIFFITEM &di = Next();

	if ((m_nFlags & DiffsOnly) && !IsItemNavigableDiff(ctxt, di))
	{
		return 0;
	}
	
	// If in Flat View simply skip over selected directories
	// This is not a complete solution; what you really want to do is inform 7z to only add the folder (not its contents) when in Flat View
	if (!GetOptionsMgr()->GetBool(OPT_TREE_MODE) && di.diffFileInfo->IsDirectory()) // mod
		return 0;

	bool isSideOnly = !di.diffcode.exists(m_index);

	...
}
```

**Comment by sdottaka** [comment]
Fixed in commit [fb882c0](https://github.com/WinMerge/winmerge/commit/fb882c0983a71c29660f79cda56990451625c269)

Thank you for the report.

**Comment by AverniteDF** [comment]
Your commit gets rid of some of the duplicates but not all of them, I fear.

In the Scripts folder, for example, where I was getting triplicates before
I am getting duplicates now. It needs to go deeper.

**Comment by AverniteDF** [comment]
Okay, I suspect it has to do with the view you are using. Your commit checks if the next item is a child of the current item, and this works well for the view in your png because children are listed immediately after their parents.

This is the view that I'm using when I get the duplications:

![completely_flat_view](https://user-images.githubusercontent.com/115361770/204343226-e3447963-c64b-4078-b868-91c21879bd9d.png)

As you can see, all the folders are listed first and then all the files last.

I believe the fix needs to work in a way that doesn't rely on the order of the listed items.

**Comment by sdottaka** [comment]
It appears you have tree mode disabled.
Your attached image shows the folder Zoom on the first line and the subfolder Out of the Zoom folder on the second line, which I would not expect to see.
I do not want to show the parent folder that has subfolders or files when tree mode is disabled.
I think this problem can be solved by not displaying this parent folder.

**Comment by sdottaka** [comment]
Fixed in commit [9b0998a](https://github.com/WinMerge/winmerge/commit/9b0998ae3342fc663f490496cc4a2673312c5104)

**Comment by AverniteDF** [comment]
Yes, this appears to be a good solution. Thank you for addressing it so quickly!
